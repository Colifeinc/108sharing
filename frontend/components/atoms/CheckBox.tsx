import * as React from 'react'
import classNames from 'classnames'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCheck } from '@fortawesome/free-solid-svg-icons/faCheck'

interface IProps {
  name: string
  defaultChecked?: boolean
  checked?: boolean
  label?: string
  value?: number | string
  onChangeHandler?(event: any): void
}

const CheckBox: React.FC<IProps> = ({
  name,
  defaultChecked,
  checked,
  label,
  onChangeHandler,
  value = 'on',
}) => {
  const hasCheckedInProp = React.useMemo(() => checked != null, [checked])
  const [checkboxChecked, setCheckboxChecked] = React.useState(defaultChecked || false)
  const actualChecked = React.useMemo(() => (hasCheckedInProp ? checked : checkboxChecked), [
    hasCheckedInProp,
    checkboxChecked,
    checked,
  ])
  const checkBoxProp = React.useMemo(
    () => ({ [hasCheckedInProp ? 'checked' : 'defaultChecked']: actualChecked }),
    [hasCheckedInProp, actualChecked]
  )

  return (
    <div className="flex">
      <div className={classNames(['flex-grow w-full flex items-center'])}>
        <label
          className={classNames([
            'relative flex items-center cursor-pointer',
            'transition duration-75 hover:opacity-85',
          ])}
        >
          <input
            className="hidden"
            type="checkbox"
            value={value}
            name={name}
            onChange={event => {
              setCheckboxChecked(event.target.checked)
              if (onChangeHandler) {
                onChangeHandler(event)
              }
            }}
            {...checkBoxProp}
          />
          <div
            className={classNames([
              'w-5 h-5 flex items-center justify-center transition duration-150 rounded-sm border-2',
              actualChecked ? 'bg-primary border-primary' : 'bg-white border-neutral-300',
            ])}
          >
            <FontAwesomeIcon
              className={classNames([
                'text-sm text-white transition duration-150',
                actualChecked ? 'opacity-100' : 'opacity-0',
              ])}
              icon={faCheck}
            />
          </div>
          <span className="ml-3 text-sm leading-snug">{label}</span>
        </label>
      </div>
    </div>
  )
}

export default CheckBox
