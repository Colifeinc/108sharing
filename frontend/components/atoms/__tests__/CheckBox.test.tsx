import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import * as React from 'react'
import CheckBox from '../CheckBox'

describe('atoms/CheckBox.tsx', () => {
  const onChangeHandler = jest.fn()
  const props = {
    name: 'test',
    defaultChecked: false,
    onChangeHandler,
  }

  it('should matches to the snapshot', () => {
    const wrapper = shallow(<CheckBox {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should call handleClick props', () => {
    const wrapper = shallow(<CheckBox {...props} />)
    wrapper.find('input').simulate('change', { target: { checked: false } })
    expect(onChangeHandler).toHaveBeenCalled()
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
