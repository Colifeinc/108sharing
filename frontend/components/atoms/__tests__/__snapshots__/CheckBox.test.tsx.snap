// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atoms/CheckBox.tsx should call handleClick props 1`] = `
<div
  className="flex"
>
  <div
    className="flex-grow w-full flex items-center"
  >
    <label
      className="relative flex items-center cursor-pointer transition duration-75 hover:opacity-85"
    >
      <input
        className="hidden"
        defaultChecked={false}
        name="test"
        onChange={[Function]}
        type="checkbox"
        value="on"
      />
      <div
        className="w-5 h-5 flex items-center justify-center transition duration-150 rounded-sm border-2 bg-white border-neutral-300"
      >
        <FontAwesomeIcon
          border={false}
          className="text-sm text-white transition duration-150 opacity-0"
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f00c",
                "M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z",
              ],
              "iconName": "check",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </div>
      <span
        className="ml-3 text-sm leading-snug"
      />
    </label>
  </div>
</div>
`;

exports[`atoms/CheckBox.tsx should matches to the snapshot 1`] = `
<div
  className="flex"
>
  <div
    className="flex-grow w-full flex items-center"
  >
    <label
      className="relative flex items-center cursor-pointer transition duration-75 hover:opacity-85"
    >
      <input
        className="hidden"
        defaultChecked={false}
        name="test"
        onChange={[Function]}
        type="checkbox"
        value="on"
      />
      <div
        className="w-5 h-5 flex items-center justify-center transition duration-150 rounded-sm border-2 bg-white border-neutral-300"
      >
        <FontAwesomeIcon
          border={false}
          className="text-sm text-white transition duration-150 opacity-0"
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f00c",
                "M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z",
              ],
              "iconName": "check",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </div>
      <span
        className="ml-3 text-sm leading-snug"
      />
    </label>
  </div>
</div>
`;
